### ISSUE TITLE
Export fastify types

### ISSUE THREAD

**Issue by TheAsda** [ts-rest/ts-rest#406]
I think types for fastify routes, options etc should be exported. This will allow developers to separate route implementations from server initialization like this:
```ts
const getPage: AppRouteImplementation<PageRoute> = async ({
  query,
  request,
}) => {
  // ...
};

const router = server.router(contracts, {
  page: getPage,
});
```

This will help organize many routes across the application.

**Comment by i-like-robots** [comment]
I was trying to figure this out in my (Express) project too and I was able to achieve it. Firstly by exporting a type for my completed contract:

```ts
// src/contract.ts
const c = initContract()

export const contract = c.router({ ... })

export type Contract = typeof contract
```

Then in the module implementing my route I import this type along with the `AppRouteImplementation` type as you showed above:

```ts
// src/routes/getThingRoute.ts
import type { Contract } from '../contract.js'
import type { AppRouteImplementation } from '@ts-rest/express'

type Route = AppRouteImplementation<Contract['getThing']>

export const getThingRoute: Route = async ({ params }) => { ... }
```

**Comment by michaelangeloio** [comment]
@TheAsda Can you please list the exact types you'd like exposed? In general, we want to limit exposing a ton of APIs (even types) to avoid breaking changes for those who might be in fact monkey-patching an API.

**Comment by TheAsda** [comment]
For now i see only `AppRouteImplementation` and `RecursiveRouterObj` would be useful. It would allow developers to create separate route implementations and combine them to nested object to fulfil the contract.

```ts
const pageRoute: AppRouteImplementation<...> = ()=>{};
const detailsRoute: AppRouteImplementation<...> = ()=>{};

const productRoutes: RecursiveRouterObj<...>={
  page: pageRoute,
  details: detailsRoute,
}

const router = server.router(contracts, {
  product: productRoutes 
})
```

**Comment by kitsunekyo** [comment]
this is not limited to fastify. i'd need this for express too:

we're trying to adopt `ts-rest` in our codebase. just like OP i want to modularize subrouters

```ts
// i get nice inferrence here
export const authRouter = s.router(contract.auth, {
  login: async ({ body: { email, password }, res }) => {}
);

//but i want to move the handler to an external file
export const authRouter = s.router(contract.auth, {
  login: loginRouter,
);
```

our route handlers look like this atm

```ts
export async function login(
  req: ICustomRequest,
  res: Response<LoginResponse>,
): Promise<void> {
// set cookies with res.cookie()
}
```

i've tried two approaches to typing the handler

 ```ts
export async function login(payload: Parameters<AppRouteImplementation<typeof contract.auth.login>>[0]) {}
```

or

```ts
export const login: AppRouteImplementation<
  typeof contract.auth.login
> = async ({ body: { email, password }, res }) => {}
```

i now get an error now "Typescript API contract, type instantiation is excessively deep and possibly infinite", which is probably what @michaelangeloio meant with monkeypatching issues.

**Comment by Gabrola** [comment]
Resolved in #537
